### ISSUE TITLE
RNN Model

### ISSUE THREAD

**Issue by Lapisbin** [Lapisbin/python-ExportAI-predictive-platform#9]

**Comment by Lapisbin** [comment]
```python
import numpy as np
import pandas as pd
import tensorflow as tf
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

# 데이터 로딩 및 전처리 함수
def load_and_preprocess_data(file_path, sequence_length=12):
    # CSV 파일 불러오기
    df = pd.read_csv('통합된_수출_예측_데이터셋.csv')
    
    # 사용할 특성(피처) 목록 정의
    features = ['원달러환율', 'GDP_성장률', '경기선행지수', '금_가격', 
                '천연가스_가격', '원유_가격', '감성지수', '수출량']
    
    # 데이터에서 해당 열만 선택하여 numpy 배열로 변환
    data = df[features].values
    
    # 데이터를 0~1 사이로 정규화
    scaler = MinMaxScaler()
    scaled_data = scaler.fit_transform(data)
    
    # 시계열 입력(X)과 타겟 출력(y) 만들기
    X, y = [], []
    for i in range(len(scaled_data) - sequence_length):
        # 연속된 시계열 데이터를 입력으로 사용
        X.append(scaled_data[i:(i + sequence_length)])
        # 다음 시점의 수출량을 타겟으로 사용
        y.append(scaled_data[i + sequence_length, -1])  # '수출량'이 마지막 열
    
    # 리스트를 numpy 배열로 변환
    X = np.array(X)
    y = np.array(y)
    
    return X, y, scaler

# RNN 모델 구성 함수 (LSTM 기반)
def create_rnn_model(sequence_length, n_features):
    model = tf.keras.Sequential([
        # 첫 번째 LSTM 층 (출력 시퀀스를 다음 LSTM에 넘김)
        tf.keras.layers.LSTM(64, activation='relu', input_shape=(sequence_length, n_features), 
                           return_sequences=True),
        # 과적합 방지를 위한 Dropout
        tf.keras.layers.Dropout(0.2),
        # 두 번째 LSTM 층 (최종 출력만 반환)
        tf.keras.layers.LSTM(32, activation='relu'),
        tf.keras.layers.Dropout(0.2),
        # Dense 층으로 출력 다듬기
        tf.keras.layers.Dense(16, activation='relu'),
        tf.keras.layers.Dense(1)  # 예측값 하나 (수출량)
    ])
    
    # 모델 컴파일 (손실 함수와 최적화 알고리즘 설정)
    model.compile(optimizer='adam', loss='mse', metrics=['mae'])
    return model

# 전체 실행 코드
def main():
    sequence_length = 12  # 12개월 데이터를 입력으로 사용
    
    # 데이터 로드 및 전처리
    X, y, scaler = load_and_preprocess_data('your_data.csv', sequence_length)
    
    # 학습/테스트 데이터 분할 (80% 학습, 20% 테스트)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    # 학습 데이터 중 일부를 검증 데이터로 분할 (학습:검증 = 64%:16%)
    X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)
    
    # 모델 생성
    model = create_rnn_model(sequence_length, X.shape[2])
    
    # EarlyStopping: 검증 성능이 향상되지 않으면 조기 종료
    early_stopping = tf.keras.callbacks.EarlyStopping(
        monitor='val_loss',      # 검증 손실 기준
        patience=10,             # 10 epoch 동안 개선 없으면 종료
        restore_best_weights=True  # 가장 좋은 성능의 가중치 복원
    )
    
    # 모델 학습
    history = model.fit(
        X_train, y_train,                # 학습 데이터
        epochs=100,                     # 최대 학습 epoch 수
        batch_size=32,                  # 배치 크기
        validation_data=(X_val, y_val), # 검증 데이터
        callbacks=[early_stopping],     # 조기 종료 콜백
        verbose=1                       # 학습 로그 출력
    )
    
    # 테스트 데이터로 모델 평가
    test_loss, test_mae = model.evaluate(X_test, y_test, verbose=0)
    print(f'\n테스트 MAE: {test_mae:.4f}')
    
    # 학습 결과 시각화 (손실, MAE)
    plt.figure(figsize=(12, 4))
    
    # 손실 시각화
    plt.subplot(1, 2, 1)
    plt.plot(history.history['loss'], label='Training Loss')
    plt.plot(history.history['val_loss'], label='Validation Loss')
    plt.title('Model Loss')
    plt.xlabel('Epoch')
    plt.ylabel('Loss')
    plt.legend()
    
    # MAE 시각화
    plt.subplot(1, 2, 2)
    plt.plot(history.history['mae'], label='Training MAE')
    plt.plot(history.history['val_mae'], label='Validation MAE')
    plt.title('Model MAE')
    plt.xlabel('Epoch')
    plt.ylabel('MAE')
    plt.legend()
    
    plt.tight_layout()
    plt.show()

# main 함수 실행
if __name__ == "__main__":
    main()
```

**Comment by Lapisbin** [comment]
```python
# Early Stopping 콜백 정의
early_stopping = tf.keras.callbacks.EarlyStopping(
    monitor='val_loss',
    patience=30,
    restore_best_weights=True
)

# Learning Rate 스케줄러 정의 
lr_scheduler = tf.keras.callbacks.ReduceLROnPlateau(
    monitor='val_loss',
    factor=0.5,
    patience=10,
    min_lr=0.00001
)

history = model.fit(X_train, y_train,
                   epochs=300,  # 에포크 수 증가
                   batch_size=16,  # 배치 사이즈 감소
                   validation_split=0.2,
                   callbacks=[early_stopping, lr_scheduler],
                   shuffle=True,  # 데이터 셔플링 추가
                   verbose=1)
```
학습 정확도 상승을 위한 조치
필요시 사용

**Comment by Lapisbin** [comment]
LSTM 모델 층 조절

**Dropout X**
```python
model = Sequential([
    LSTM(64, activation='relu', input_shape=(X_train.shape[1], X_train.shape[2]), return_sequences=True),
    LSTM(32, activation='relu'),
    Dense(16, activation='relu'),
    Dense(1)
])
```

**Dropout O**
```python
# 모델 정의
model = Sequential([
    LSTM(64, activation='relu', input_shape=(X_train.shape[1], X_train.shape[2]), return_sequences=True),
    Dropout(0.2),
    LSTM(32, activation='relu'),
    Dropout(0.2),
    Dense(16, activation='relu'),
    Dense(1)
])
```
